engine/server.go

```diff
@@ -0,0 +1,97 @@
+package engine
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"time"
+)
+
+type LogLevel int
+
+const (
+	LogDebug LogLevel = iota
+	LogInfo
+	LogWarn
+	LogError
+)
+
+type Logger interface {
+	Errorf(fmt string, v ...interface{})
+	Warnf(fmt string, v ...interface{})
+	Infof(fmt string, v ...interface{})
+	Debugf(fmt string, v ...interface{})
+}
+
+type Options func(*Server)
+
+func OptionPingInterval(time.Duration) Options                      { return nil }
+func OptionPingTimeout(time.Duration) Options                       { return nil }
+func OptionMaxBufferSize(int) Options                               { return nil }
+func OptionLogLevel(level LogLevel) Options                         { return nil }
+func OptionLogger(logger Logger) Options                            { return nil }
+func OptionTransports(initial string, upgradings ...string) Options { return nil }
+func OptionJSONP(padding int) Options                               { return nil }
+
+type FrameType int
+
+const (
+	FrameBinary FrameType = iota
+	FrameText
+)
+
+type PacketType int
+
+const (
+	PacketOpen PacketType = iota
+	PacketClose
+	PacketPing
+	PacketPong
+	PacketMessage
+	PacketUpgrade
+	PacketNoop
+)
+
+type Server struct{}
+
+func New(...Options) (*Server, error) {
+	return nil, nil
+}
+
+// OnXXX should be called before serving HTTP.
+// The engineio framework processes next messages after OnXXX() done. All callback passing to OnXXX should return ASAP.
+func (s *Server) OnOpen(func(*Context, *http.Request) error) {}
+func (s *Server) OnMessage(func(*Context, io.Reader))        {}
```

Comment:
For #374, I'm thinking about the usage of context.

From the concept, a context should be meta info of a chain of RPC calls. In socket.io, it should not equal the context in HTTP requests, especially the case of using WebSocket as a transport.

Besides that, socket.io doesn't have a way to put meta info when calling an RPC. There is no good way to pass a deadline or values(like RPC tracking id) without breaking into the value.

As reasons above, the root context may always be a `context.Background()`, which is meaningless.

There is a way of passing meta info: putting info in the input of RPC, turning them as context by middleware, and injecting into another input by another middleware when calling other RPCs. The problem is, it sounds very complex.

Reply:
Why we couldn't use `context.Context` value as global shared context for throw some values? i saw code that use this exception in production code. for example, global logger push through from middlewares into context

Reply:
Never mind, I think it still worth keeping this design.